### ISSUE TITLE
[Feature] Generalized Mean Pooling

### ISSUE THREAD

**Issue by Atharva-Phatak** [catalyst-team/catalyst#1084]
## 🚀 Feature Request
I would catalyst to have an implementation of GeM(Generalized Mean Pooling) as proposed in this [paper](https://arxiv.org/pdf/1711.02512.pdf)

### Motivation
Well I like to use catalyst's contrib features very much and it has an amazing set of pooling methods available(GlobalMaxPool2d , 
GlobalConcatPool2d ,etc). I think that an implementation of GeM pooling will be a nice addition to the pooling module of catalyst.

## Note
Goes without saying that I will work on this implementation myself

**Comment by Scitator** [comment]
Hi, 

The idea with the contrib pooling looks great for me (especially for our representation learning track). In this case, if you have time and motivation – it would be a valuable contribution. As far as there is already an implementation from the authors, I hope it would be a bit easier to add to the contrib. Nevertheless, ensure please that you have added docs and a few tests for new functionality. I know, that is not the most interesting part of the contribution, but the most long-term one for the user adoption.

**Comment by Atharva-Phatak** [comment]
Definetly, I will create a pull request including documentation and the required tests along with accepted implementations

Also since you mentioned representation learning for catalyst, would like to add more modules/Loss functions ? Like arcface , cosface, etc?

**Comment by Scitator** [comment]
The interesting thing... we already have many [arcfaces implemented and tested](https://medium.com/catalyst-team/representation-learning-with-catalyst-and-faces-946644d49184). The same for [metric learning](https://medium.com/pytorch/metric-learning-with-catalyst-8c8337dfab1a).

That's why all modules/losses/approaches for representation and metric learning are always welcome in the Catalyst ;)
